torch/_refs/__init__.py

```diff
@@ -3459,6 +3460,66 @@ def meshgrid(
     return grids
 
 
+# CompositeImplicitAutograd - don't register decomp
+def movedim(
+    input: TensorLikeType,
+    source: Union[int, DimsSequenceType],
+    destination: Union[int, DimsSequenceType],
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.movedim
+    """
+    if type(source) is int:
+        source = (source,)
+    if type(destination) is int:
+        destination = (destination,)
+
+    utils.check(
+        len(source) == len(destination),  # type: ignore[arg-type]
+        lambda: (
+            "movedim: Invalid source or destination dims: source "
+            f"({source} dims) should contain the same number of dims as "
+            f"destination ({destination} dims)"
+        ),
+    )
+
+    rank = input.ndim
+    ss = [utils.canonicalize_dim(rank=rank, idx=idx) for idx in source]  # type: ignore[union-attr]
+    ds = [utils.canonicalize_dim(rank=rank, idx=idx) for idx in destination]  # type: ignore[union-attr]
+
+    sss = set(ss)
+    dss = set(ds)
+
+    utils.check(
+        len(ss) == len(sss),
+        lambda: f"movedim: repeated dim in `source` {source}",
+    )
+    utils.check(
+        len(ds) == len(dss),
+        lambda: f"movedim: repeated dim in `destination` {destination}",
+    )
+
+    m = dict(zip(ds, ss))
+    dims = []
+    si = 0  # source index
+    for di in range(rank):
+        # check if the destination index is in the mapping
+        s = m.get(di)
+        if s is not None:
+            # insert source index if found
+            dims.append(s)
+        else:
+            # insert source index sequentially, skipping indices from the mapping
+            while si in sss:
```

Comment:
why are you iterating over `sss` and not `ss`?

Reply:
this tests for membership, which is a constant operation on sets, so it should be faster in theory. for small collections it shouldn't matter much, but i have to create `sss` anyway to check for duplicates earlier, so i'm using it here as well